source/grid/grid_tools.cc

```diff
@@ -414,6 +415,76 @@ namespace GridTools
   }
 
 
+  template <int dim>
+  Vector<double>
+  calculate_aspect_ratio_of_cells(const Triangulation<dim> &triangulation,
+                                  const Mapping<dim> &      mapping,
+                                  const Quadrature<dim> &   quadrature)
+  {
+    FE_Nothing<dim> fe;
+    FEValues<dim>   fe_values(mapping, fe, quadrature, update_jacobians);
+
+    Vector<double> aspect_ratio_vector(triangulation.n_active_cells());
+
+    // loop over cells of processor
+    for (const auto &cell : triangulation.active_cell_iterators())
+      {
```

Comment:
I think you might want to skip artificial cells (does this even work right now?).

Reply:
Sorry, I need some background. Are you happy with the previous (non range-based) version, or is it the same for both versions?

Reply:
I don't know what an ``artificial`` cell is. I agree with ``locally_owned``. What I found in the deal.II docu is that ``locally_owned`` and ``!artificial`` is clearly something different, so I am not sure if I really got your point (assume that this is just an example? What info does a processor have regarding ghost cells vs. artificial cells? Do you want to imply that the computation would be correct for a ghost cell?).

Reply:
changed in https://github.com/dealii/dealii/pull/8894/commits/b9ae00282d0a425fbcd7b8c6dc37af4bb40950d1

Reply:
The glossary (https://www.dealii.org/current/doxygen/deal.II/DEALGlossary.html) says:
> Artificial cells
    If a mesh is distributed across multiple MPI processes using the parallel::distributed::Triangulation class, each processor stores only the cells it owns, one layer of adjacent cells that are owned by other processors (called ghost cells), all coarse level cells, and all cells that are necessary to maintain the invariant that adjacent cells must differ by at most one refinement level. The cells stored on each process that are not owned by this process and that are not ghost cells are called "artificial cells", and for these cells the predicate cell->is_artificial() returns true. Artificial cells are guaranteed to exist in the globally distributed mesh but they may be further refined on other processors. See the Distributed Computing paper for more information.
    The concept of artificial cells has no meaning for triangulations that store the entire mesh on each processor, i.e. the Triangulation class. 

Artifical cells don't necessarily carry the correct, i.e. globall consistent, geometric informations while this is true for locally owned and ghosted cells.

Reply:
> Do you want to imply that the computation would be correct for a ghost cell?).

Likely, unless one processor decides to move vertices of only their own cells around.

Reply:
I would consider wrong information for ghost cells a bug - DG methods taking derivatives on faces critically depend on the ghost cells to be consistent. (Hyperbolic terms should be fine if all vertices to the locally owned cells are consistent across processors because the normal vector and integration factors do only depend on the vertices on the face; whereas gradients need *all* vertices of a ghost cell.)

Nonetheless, I agree with @tjhei that we should only use the own cells - this is assumed by `VectorTools::compute_global_error` which is used here. Also, we should initialize the vector with `0` so artificial cells always get that value. The typical use case is to compute a max over all owned cells (where only owned cells are used) or to pass that vector to `DataOut::add_data_vector` where one also ignores non-owned cells, so the state is correct. And an aspect ratio of 0 signals that nothing has happened as the definition cannot be less than 1.

Reply:
Initialization with 0 is done by the constructor, isn't it?

Reply:
Yes.